### ISSUE TITLE
Image parsing 

### ISSUE THREAD

**Issue by santosh9sanjeev** [minar09/cp-vton-plus#15]
I am trying to give inputs from internet like take an image of a person and an image of the cloth..
I tried running the dataset_neck_skin_connection.py
How should I get the image-parse for the input image.. pls guide me.. I am a beginner

**Comment by minar09** [comment]
Hi @santosh9sanjeev , to run the model with custom internet images, make sure you have the following:

1. image (image of a person)
2. image-parse (you can generate with LIP_JPPNet or CIHP_PGN pretrained networks from the person image. Then you can run dataset_neck_skin_connection.py for LIP parsing. Finally run body_binary_masking.py)
3. cloth (in-shop cloth image)
4. cloth-mask (binary mask of cloth image, you can generate it with simple pillow/opencv function)
5. pose (pose keypoints of the person, generate with openpose COCO 18 model )

Hope that helps. Thank you.

**Comment by santosh9sanjeev** [comment]
Thank you @minar09 for your help 
I completed all the steps as you mentioned but I am facing problems with bit depth. The image-parse generated by CIHP_PGN is generating an image of bit depth 24 whereas the one needed is bit depth is 8. Using pillow I converted the bit depth from 24 to 8 but then the image mask generated had weird result... It actually did not correctly generate the mask..
Any idea how to proceed or to solve this error.. It would be very helpful..

**Comment by thaithanhtuan** [comment]
> Thank you @minar09 for your help
> I completed all the steps as you mentioned but I am facing problems with bit depth. The image-parse generated by CIHP_PGN is generating an image of bit depth 24 whereas the one needed is bit depth is 8. Using pillow I converted the bit depth from 24 to 8 but then the image mask generated had weird result... It actually did not correctly generate the mask..
> Any idea how to proceed or to solve this error.. It would be very helpful..

It is not clear how weird result for the mask generated. Can you upload or share the code for cp_dataset or the result of CIHP_PGN. May be the file format is different from VITON dataset.

**Comment by santosh9sanjeev** [comment]
![person-final-1](https://user-images.githubusercontent.com/51113988/91655164-6233a080-eacc-11ea-9715-139d4b4deba8.png)
![person-final-2](https://user-images.githubusercontent.com/51113988/91655165-6364cd80-eacc-11ea-9655-9315d70a15a5.png)
Both the files are of png format, having dimension 192,256 and bit depth 8
When i ran the code for these images i got the following masked images
![person-final-2](https://user-images.githubusercontent.com/51113988/91655234-c9515500-eacc-11ea-8546-42a8e5c427a6.png)
![person-final-1](https://user-images.githubusercontent.com/51113988/91655236-cb1b1880-eacc-11ea-833c-6f2f7436d5e5.png)

**Comment by minar09** [comment]
Hi @santosh9sanjeev , please make sure to use grayscale files as input, not the RGB files from the CIHP_PGN generated segmentation. Also, I think you don't need to create binary shape masks directly before testing for CIHP_PGN (although there is no harm in that, creating binary masks script was made for VITON/LIP style segmentation), since CIHP_PGN produces better segmentation with torso-neck label. You can get the body shape for input in `cp_dataset.py` as following code:
        `im_parse = Image.open(osp.join(self.data_path, 'image-parse', parse_name))
        parse_array = np.array(im_parse)
        parse_shape = (parse_array > 0).astype(np.float32)`

**Comment by santosh9sanjeev** [comment]
Thank you so much minar09 for your help...
I just have another small doubt actually the test.py code requires image-parse-new and image-mask folder..
should i change the code where there is image-mask and image-parse-new to image-parse, as I am not generating the image-mask and image-parse-new, or is ther any better alternative?
Thank you once again for your help..

**Comment by minar09** [comment]
> Thank you so much minar09 for your help...
> I just have another small doubt actually the test.py code requires image-parse-new and image-mask folder..
> should i change the code where there is image-mask and image-parse-new to image-parse, as I am not generating the image-mask and image-parse-new, or is ther any better alternative?
> Thank you once again for your help..

Yes, just change the code if you don't need them. Only updating in `cp_dataset.py` should be enough.

**Comment by minar09** [comment]
Closing the issue as its resolved. Feel free to reopen in case there are still problems. Results with custom images: #23

**Comment by Pritam-N** [comment]
> Hi @santosh9sanjeev , please make sure to use grayscale ([0, 20]) files as input, not the RGB files from the CIHP_PGN generated segmentation. Also, I think you don't need to create binary shape masks directly before testing for CIHP_PGN (although there is no harm in that, creating binary masks script was made for VITON/LIP style segmentation), since CIHP_PGN produces better segmentation with torso-neck label. You can get the body shape for input in `cp_dataset.py` as following code:
> 
> `im_parse = Image.open(osp.join(self.data_path, 'image-parse', parse_name)) # read segmentation`
> `parse_array = np.array(im_parse) # convert to numpy array`
> `parse_shape = (parse_array > 0).astype(np.float32) # get binary body shape`

I am trying with segmentation generated from Graphonomy. Any suggestion on how to convert that to grayscale with 0-20 classes?

**Comment by minar09** [comment]
Actually, any image segmentation network should originally generate grayscale output, so please check the actual network output.

**Comment by yashp0103** [comment]
Hi @santosh9sanjeev @minar09 

Thanks for the information!
I'm able to test on custom image with some manual work on JSON keypoints. The openpose generated JSON file has two different fields like face_keypoints_2d and face_keypoints_3d. And In cp_vton_plus model JSON contains field like face_keypoints only. So, I just want to confirm that the JSONs present in model are manually modified or directly generated from openpose? 

If it is directly generated then can you help me to understand what changes you had done or what run cmd you had used.

Thanks,
Yash

**Comment by minar09** [comment]
@yashp0103 , CP-VTON+ model directly uses openpose-generated keypoints, no modification is needed. And the face_keypoints are not used in this model, you can ignore them. If your joints have pose_keypoints_2d, you can directly use it by just changing this line: https://github.com/minar09/cp-vton-plus/blob/master/cp_dataset.py#L152 to ```pose_data = pose_label['people'][0]['pose_keypoints_2d']
```

**Comment by yashp0103** [comment]
Thank you so much @minar09 
It's working!

**Comment by Amin-asadii** [comment]
Please send the link CIHP_PGN.
thank you

**Comment by Amin-asadii** [comment]
Thank you so much minar09 for your help...
I use LIP_JPPNet for image-pars.But the results are not good at all. Please see for this image:
![000001_01](https://user-images.githubusercontent.com/80091528/132836451-8a448900-5d69-4bfa-9970-b411787cbffd.jpg)
The following results are obtained:
![000001_0_vis](https://user-images.githubusercontent.com/80091528/132836596-68ff0abe-4e98-4019-9e28-9d64d2837590.png)
![000001_0](https://user-images.githubusercontent.com/80091528/132836661-e071d61a-fac3-4380-98da-2e852d5454c3.png)
If I was expecting this output:
![000001_12 (1)](https://user-images.githubusercontent.com/80091528/132837003-62033267-7f46-42d2-9681-1fa0b9c47c66.png)
![000001_12 (2)](https://user-images.githubusercontent.com/80091528/132837040-a79586c1-5544-414d-95f9-6065120d65dd.png)
Thank you very much for your help:
Thank

**Comment by Amin-asadii** [comment]
00989174286532 Whats App

**Comment by minar09** [comment]
@Amin-asadii , you can try CIHP-PGN pre-trained model for parsing, which should give better results.

**Comment by Amin-asadii** [comment]
Python
91.0%
 
MATLAB
9.0%

 
Owner
@minar09  Do I have to have matlab installed?
How should I use it?

**Comment by minar09** [comment]
@Amin-asadii , no need to install Matlab for CP-VTON+.

**Comment by Amin-asadii** [comment]
@minar09 I mean to use CIHP-PGN

**Comment by Amin-asadii** [comment]
Hello my dear friend
Thank you very much for answering our questions with love.
I need to run CP-VTON + for a research project.
But I am not able to execute CIHP_PGN and openpose COCO-18 codes.
Please advise if the code is running in the colab or another way I can use the above keys.
thanks a lot
WhatsApp 00989174286232

**Comment by Amin-asadii** [comment]
@minar09 Hello dear friend, thank you for your help
Please attach the openpose COCO-18 Python link

**Comment by minar09** [comment]
@Amin-asadii , here you go: https://github.com/CMU-Perceptual-Computing-Lab/openpose/releases/download/v1.7.0/openpose-1.7.0-binaries-win64-gpu-python3.7-flir-3d_recommended.zip

**Comment by Amin-asadii** [comment]
tank you for help you

‫‪Matiur Rahman Minar‬‏ ***@***.***‬‏> در تاریخ شنبه ۱۸
سپتامبر ۲۰۲۱ ساعت ۹:۰۴ نوشت:‬

**Comment by shu0o0yX** [comment]
> Thank you so much minar09 for your help... I use LIP_JPPNet for image-pars.But the results are not good at all. Please see for this image: ![000001_01](https://user-images.githubusercontent.com/80091528/132836451-8a448900-5d69-4bfa-9970-b411787cbffd.jpg) The following results are obtained: ![000001_0_vis](https://user-images.githubusercontent.com/80091528/132836596-68ff0abe-4e98-4019-9e28-9d64d2837590.png) ![000001_0](https://user-images.githubusercontent.com/80091528/132836661-e071d61a-fac3-4380-98da-2e852d5454c3.png) If I was expecting this output: ![000001_12 (1)](https://user-images.githubusercontent.com/80091528/132837003-62033267-7f46-42d2-9681-1fa0b9c47c66.png) ![000001_12 (2)](https://user-images.githubusercontent.com/80091528/132837040-a79586c1-5544-414d-95f9-6065120d65dd.png) Thank you very much for your help: Thank

 Could you please share the code? I'm doing this work, but I don't know how to get two pictures like these (image-parse and image-parse-new) by code, Thanks!
